### ISSUE TITLE
Security: Source code and trust

### ISSUE THREAD

**Issue by naim94a** [zed-industries/zed#7455]
### Check for existing issues

- [X] Completed

### Describe the feature

Zed is amazing when working on my own projects, or opening projects that I trust. 
I think it would be good to prompt the user if they trust the source code before starting any process that may cause the user to execute potentially malicious code, (e.g. a malicious `build.rs`). 
Perhaps also allow the user to specify if a directory is trusted or untrusted by default.

**Comment by Bloeckchengrafik** [comment]
I think this is a good idea! This could also be integrated with [git safe directories](https://git-scm.com/docs/git-config#Documentation/git-config.txt-safedirectory) config.

**Comment by jbms** [comment]
In addition to unsafe language servers, external commands (e.g. for formatting) specified in the project config is another issue.

**Comment by MrSubidubi** [comment]
Yeah that definitely seems like a reasonable request. We did look into this a bit already, but seems worth pursuing more. Thanks!

**Comment by szabgab** [comment]
When I launch zed now on a new project it is in "Restricted mode" mentioned in the top left corners. Isn't that the implementation of this request?

**Comment by swannysec** [comment]
> When I launch zed now on a new project it is in "Restricted mode" mentioned in the top left corners. Isn't that the implementation of this request?

While that was my first instinct, I think there's some interesting nuance here (which is relevant to future plans in this space).  Here's the docs on our current worktree trust function: https://zed.dev/docs/worktree-trust

Here's the key detail (bold is my emphasis):

> Restricted Mode prevents downloading and running related items from .zed/settings.json. Until a worktree is trusted, Zed does not run related untrusted actions and waits for user confirmation. This gives users a chance to review project settings, MCP servers, and language servers.
>
> **Zed still trusts tools it installs globally. Global MCP servers and global language servers such as Prettier and Copilot are installed and started as usual, independent of worktree trust.**

@naim94a - can you confirm whether the existing worktree trust function is what you're looking for or whether the bold portion above is the ask, in addition to the existing feature?

If the latter, while I can't give any estimates on timeline, it's on my list to consider the expansion of worktree trust to account for other executable surfaces.  This is a more complex task than it seems on the surface, but know that it's important to us to provide you all with the tools to manage your own level of risk tolerance and threat model.
